### PR TITLE
Update to jupyterlab 4 and lumino 2

### LINKED ISSUE #3700: Available for documentation/information about ipywidgets and Jupyterlab 4
Hello, I am opening this issue with some questions about using Jupyterlab 4 and ipywidgets.
I am wondering if as of now there any outlined steps that widget maintainers will have to take? I saw that previously, a section in the [changelog](https://ipywidgets.readthedocs.io/en/stable/changelog.html#updates-for-widget-maintainers) has been included with details about any steps widget maintainers needed to take for the update from Lab 2 to Lab 3. It seems that these helpful instructions were provided by @jasongrout.. I am wondering if there is any similar documentation that one might reference as we begin trying out the latest releases of Jupyterlab 4?

### PR DESCRIPTION
**Pull request by echarles**
Fixes https://github.com/jupyter-widgets/ipywidgets/issues/3700

### CODE REVIEW

File: packages/base/package.json [hunk 1]
```diff
@@ -33,10 +33,10 @@
...
   },
   "dependencies": {
-    "@jupyterlab/services": "^6.0.0",
-    "@lumino/coreutils": "^1.11.1",
-    "@lumino/messaging": "^1.10.1",
-    "@lumino/widgets": "^1.30.0",
+    "@jupyterlab/services": "^7.0.0-alpha.19",
+    "@lumino/coreutils": "^2.0.0-beta.0",
+    "@lumino/messaging": "^2.0.0-beta.0",
+    "@lumino/widgets": "^2.0.0-beta.0",
```
Comment (martinRenou) [inline packages/base/package.json:39]: I'd suggest being soft on the Lumino dependencies:

```suggestion
    "@lumino/coreutils": "^1.11.1 || ^2.0.0-beta.0",
    "@lumino/messaging": "^1.10.1 || ^2.0.0-beta.0",
    "@lumino/widgets": "^1.30.0 || ^2.0.0-beta.0",
```

That way people won't get warnings in the JS console with JupyterLab 3.

**Review by martinRenou — COMMENTED** [review]
Thanks for starting this!

File: packages/base/package.json (hunk 1 above)
Comment (echarles) [reply packages/base/package.json:39]: yes, that would be the goal. I'd like first to get the CI green with more strict constraints on lumino.

**Comment by echarles** [comment]
Among the failed CI jobs, there is the Javascript one. Is CI main branch supposed to be 100% green?

```
@jupyter-widgets/html-manager: webpack 5.75.0 compiled with 3 warnings in 20474 ms
@jupyter-widgets/html-manager: > @jupyter-widgets/html-manager@1.0.5 build:embed-amd
@jupyter-widgets/html-manager: > node scripts/concat-amd-build.js && rimraf dist/amd


 >  Lerna (powered by Nx)   Running target build for 7 projects failed

   Tasks not run because their dependencies failed or --nx-bail=true:
   
   - @jupyter-widgets/notebook-manager:build
   
   Failed tasks:
   
   - @jupyter-widgets/jupyterlab-manager:build

error Command failed with exit code 1.
info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.
Error: Process completed with exit code 1.
```

**Comment by echarles** [comment]
@martinRenou I have pushed fixes and this PR still requests approval. Is this possible to simply have the CI run on each commit without asking approval?

**Comment by echarles** [comment]
I have pushed more updates for the CI and the CI running in my fork is now better. There is still one failure related to galata usage, see the details here. @fcollonval may have an idea how to investigate this, probably related to galata API changes between 3.x and 4.x ?

```
cd ui-tests
  yarn test
  shell: /usr/bin/bash -e {0}
  env:
    pythonLocation: /opt/hostedtoolcache/Python/[3](https://github.com/datalayer-externals/ipywidgets/actions/runs/4269369026/jobs/7432505017#step:11:3).11.2/x6[4](https://github.com/datalayer-externals/ipywidgets/actions/runs/4269369026/jobs/7432505017#step:11:4)
    PKG_CONFIG_PATH: /opt/hostedtoolcache/Python/3.11.2/x64/lib/pkgconfig
    Python_ROOT_DIR: /opt/hostedtoolcache/Python/3.11.2/x64
    Python2_ROOT_DIR: /opt/hostedtoolcache/Python/3.11.2/x64
    Python3_ROOT_DIR: /opt/hostedtoolcache/Python/3.11.2/x64
    LD_LIBRARY_PATH: /opt/hostedtoolcache/Python/3.11.2/x64/lib
  
yarn run v1.22.19
$ playwright test

Running 1 test using 1 worker
×F

  1) tests/widgets.test.ts:19:7 › Widget Visual Regression › Run notebook widgets.ipynb and capture cell outputs 

    Error: Failed to activate galata extension

        at Proxy.hookHelpersUp (/home/runner/work/ipywidgets/ipywidgets/ui-tests/node_modules/@jupyterlab/galata/src/jupyterlabpage.ts:669:13)
        at Proxy.goto (/home/runner/work/ipywidgets/ipywidgets/ui-tests/node_modules/@jupyterlab/galata/src/jupyterlabpage.ts:474:[5](https://github.com/datalayer-externals/ipywidgets/actions/runs/4269369026/jobs/7432505017#step:11:5))
        at Object.initTestPage (/home/runner/work/ipywidgets/ipywidgets/ui-tests/node_modules/@jupyterlab/galata/src/galata.ts:284:7)
        at Object.page [as fn] (/home/runner/work/ipywidgets/ipywidgets/ui-tests/node_modules/@jupyterlab/galata/src/fixtures.ts:371:7)

    attachment #1: video (video/webm) ──────────────────────────────────────────────────────────────
    test-results/tests-widgets-Widget-Visual-Regression-Run-notebook-widgets-ipynb-and-capture-cell-outputs/video.webm
    ────────────────────────────────────────────────────────────────────────────────────────────────

    Retry #1 ───────────────────────────────────────────────────────────────────────────────────────

    Error: Failed to activate galata extension

        at Proxy.hookHelpersUp (/home/runner/work/ipywidgets/ipywidgets/ui-tests/node_modules/@jupyterlab/galata/src/jupyterlabpage.ts:[6](https://github.com/datalayer-externals/ipywidgets/actions/runs/4269369026/jobs/7432505017#step:11:6)69:13)
        at Proxy.goto (/home/runner/work/ipywidgets/ipywidgets/ui-tests/node_modules/@jupyterlab/galata/src/jupyterlabpage.ts:4[7](https://github.com/datalayer-externals/ipywidgets/actions/runs/4269369026/jobs/7432505017#step:11:7)4:5)
        at Object.initTestPage (/home/runner/work/ipywidgets/ipywidgets/ui-tests/node_modules/@jupyterlab/galata/src/galata.ts:2[8](https://github.com/datalayer-externals/ipywidgets/actions/runs/4269369026/jobs/7432505017#step:11:8)4:7)
        at Object.page [as fn] (/home/runner/work/ipywidgets/ipywidgets/ui-tests/node_modules/@jupyterlab/galata/src/fixtures.ts:371:7)

    attachment #1: video (video/webm) ──────────────────────────────────────────────────────────────
    test-results/tests-widgets-Widget-Visual-Regression-Run-notebook-widgets-ipynb-and-capture-cell-outputs-retry1/video.webm
    ────────────────────────────────────────────────────────────────────────────────────────────────

  1 failed
    tests/widgets.test.ts:1[9](https://github.com/datalayer-externals/ipywidgets/actions/runs/4269369026/jobs/7432505017#step:11:9):7 › Widget Visual Regression › Run notebook widgets.ipynb and capture cell outputs 
Error:   1) tests/widgets.test.ts:19:7 › Widget Visual Regression › Run notebook widgets.ipynb and capture cell outputs 
    Error: Failed to activate galata extension

        at Proxy.hookHelpersUp (/home/runner/work/ipywidgets/ipywidgets/ui-tests/node_modules/@jupyterlab/galata/src/jupyterlabpage.ts:669:[13](https://github.com/datalayer-externals/ipywidgets/actions/runs/4269369026/jobs/7432505017#step:11:14))
        at Proxy.goto (/home/runner/work/ipywidgets/ipywidgets/ui-tests/node_modules/@jupyterlab/galata/src/jupyterlabpage.ts:474:5)
        at Object.initTestPage (/home/runner/work/ipywidgets/ipywidgets/ui-tests/node_modules/@jupyterlab/galata/src/galata.ts:284:7)
        at Object.page [as fn] (/home/runner/work/ipywidgets/ipywidgets/ui-tests/node_modules/@jupyterlab/galata/src/fixtures.ts:371:7)
Error:   1) tests/widgets.test.ts:[19](https://github.com/datalayer-externals/ipywidgets/actions/runs/4269369026/jobs/7432505017#step:11:20):7 › Widget Visual Regression › Run notebook widgets.ipynb and capture cell outputs 

    Retry #1 ───────────────────────────────────────────────────────────────────────────────────────
    Error: Failed to activate galata extension

        at Proxy.hookHelpersUp (/home/runner/work/ipywidgets/ipywidgets/ui-tests/node_modules/@jupyterlab/galata/src/jupyterlabpage.ts:669:13)
        at Proxy.goto (/home/runner/work/ipywidgets/ipywidgets/ui-tests/node_modules/@jupyterlab/galata/src/jupyterlabpage.ts:474:5)
        at Object.initTestPage (/home/runner/work/ipywidgets/ipywidgets/ui-tests/node_modules/@jupyterlab/galata/src/galata.ts:284:7)
        at Object.page [as fn] (/home/runner/work/ipywidgets/ipywidgets/ui-tests/node_modules/@jupyterlab/galata/src/fixtures.ts:371:7)
Notice:   1 failed
    tests/widgets.test.ts:19:7 › Widget Visual Regression › Run notebook widgets.ipynb and capture cell outputs 
error Command failed with exit code 1.
info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.
Error: Process completed with exit code 1.
```

**Comment by fcollonval** [comment]
@echarles the setup of galata in JLab 4 is changing (the helpers are loaded as a regular JupyterLab extension):

https://github.com/jupyterlab/jupyterlab/tree/master/galata#launch-jupyterlab

**Comment by echarles** [comment]
Thx for the link. I have tried to implement those changes in a simple extension without luck. Do you have a link to an extension depending on jupyterlab4 that runs galata tests?

**Comment by jtpio** [comment]
> Thx for the link. I have tried to implement those changes in a simple extension without luck. Do you have a link to an extension depending on jupyterlab4 that runs galata tests?

Maybe this PR that updated the cookiecutter? https://github.com/jupyterlab/extension-cookiecutter-ts/pull/287

**Comment by echarles** [comment]
Suprised to see https://github.com/jupyter-widgets/ipywidgets/pull/3752.

Can you sync?

**Comment by tarrade** [comment]
Jupyterlab 4.0.0 was release yesterday https://github.com/jupyterlab/jupyterlab/releases/tag/v4.0.0 and it seems a new version with this PR is needed to work with jupyterlab-4.0.0, correct ?

Using conda, I am getting:
"package jupyterlab_widgets-3.0.7-pyhd8ed1ab_0 has constraint jupyterlab >=3,<4 conflicting with jupyterlab-4.0.0-pyhd8ed1ab_1"

**Comment by martinRenou** [comment]
> Using conda, I am getting:
"package jupyterlab_widgets-3.0.7-pyhd8ed1ab_0 has constraint jupyterlab >=3,<4 conflicting with jupyterlab-4.0.0-pyhd8ed1ab_1"

@tarrade this has to do with the conda package itself which has a run constraint. It was resolved in https://github.com/conda-forge/jupyterlab_widgets-feedstock/pull/26 and you will be able to install latest ipywidgets with JupyterLab 4 in ~1 hour.

**Comment by tarrade** [comment]
> > Using conda, I am getting:
> > "package jupyterlab_widgets-3.0.7-pyhd8ed1ab_0 has constraint jupyterlab >=3,<4 conflicting with jupyterlab-4.0.0-pyhd8ed1ab_1"
> 
> @tarrade this has to do with the conda package itself which has a run constraint. It was resolved in [conda-forge/jupyterlab_widgets-feedstock#26](https://github.com/conda-forge/jupyterlab_widgets-feedstock/pull/26) and you will be able to install latest ipywidgets with JupyterLab 4 in ~1 hour.

perfect, pb solved, thanks a lot

**Pull request closed without merging**